File: datasets/Python/unpaired/AI/levenshtein_1.py
def levenshtein_distance(str1, str2):
    """
    Calculate the Levenshtein distance between two strings.
    """
    # Size of the matrix
    rows = len(str1) + 1
    cols = len(str2) + 1

    # Initialize the matrix
    dist = [[0 for _ in range(cols)] for _ in range(rows)]

    # Populate the matrix
    for i in range(1, rows):
        dist[i][0] = i
    for j in range(1, cols):
        dist[0][j] = j

    # Calculate distances
    for i in range(1, rows):
        for j in range(1, cols):
            if str1[i - 1] == str2[j - 1]:
                cost = 0
            else:
                cost = 1

            dist[i][j] = min(dist[i - 1][j] + 1,      # Deletion
                             dist[i][j - 1] + 1,      # Insertion
                             dist[i - 1][j - 1] + cost)  # Substitution

    return dist[-1][-1]

# Example usage
levenshtein_example = levenshtein_distance("kitten", "sitting")
levenshtein_example  # This should return the Levenshtein distance between "kitten" and "sitting"

File: datasets/go/train/AI/levenshtein1.go
package main

import (
	"fmt"
)

// levenshteinIterative calculates the Levenshtein distance between two strings iteratively.
func levenshteinIterative(str1, str2 string) int {
	lenStr1 := len(str1)
	lenStr2 := len(str2)
	dp := make([][]int, lenStr1+1)
	for i := range dp {
		dp[i] = make([]int, lenStr2+1)
	}

	for i := 0; i <= lenStr1; i++ {
		dp[i][0] = i
	}
	for j := 0; j <= lenStr2; j++ {
		dp[0][j] = j
	}

	for i := 1; i <= lenStr1; i++ {
		for j := 1; j <= lenStr2; j++ {
			cost := 0
			if str1[i-1] != str2[j-1] {
				cost = 1
			}
			dp[i][j] = min(
				dp[i-1][j]+1,   // Deletion
				min(
					dp[i][j-1]+1,       // Insertion
					dp[i-1][j-1]+cost, // Substitution
				),
			)
		}
	}
	return dp[lenStr1][lenStr2]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	str1 := "kitten"
	str2 := "sitting"
	fmt.Printf("Levenshtein distance between '%s' and '%s' is %d\n", str1, str2, levenshteinIterative(str1, str2))
}
